src/app/server/CommissioningWindowManager.cpp

```diff
@@ -0,0 +1,277 @@
+{
+    mServer->GetExchangManager().UnregisterUnsolicitedMessageHandlerForType(Protocols::SecureChannel::MsgType::PBKDFParamRequest);
+    StopAdvertisement();
+
+    // reset all advertising
+    app::MdnsServer::Instance().StartServer(Mdns::CommissioningMode::kDisabled);
+}
+
+void CommissioningWindowManager::OnSessionEstablishmentError(CHIP_ERROR err)
+{
+    Cleanup();
+
+    ChipLogProgress(AppServer, "OnSessionEstablishmentError: %s", ErrorStr(err));
+    ChipLogProgress(AppServer, "Failed in SPAKE2+ handshake");
+
+    if (mAppDelegate != nullptr)
+    {
+        mAppDelegate->OnRendezvousStopped();
+    }
+}
+
+void CommissioningWindowManager::OnSessionEstablished()
+{
+    CHIP_ERROR err = mServer->GetSecureSessionManager().NewPairing(
+        Optional<Transport::PeerAddress>::Value(mPairingSession.GetPeerAddress()), mPairingSession.GetPeerNodeId(),
+        &mPairingSession, SecureSession::SessionRole::kResponder, 0);
+    if (err != CHIP_NO_ERROR)
+    {
+        ChipLogError(AppServer, "Failed in setting up secure channel: err %s", ErrorStr(err));
+        OnSessionEstablishmentError(err);
+        return;
+    }
+
+    ChipLogProgress(AppServer, "Device completed SPAKE2+ handshake");
+    if (mAppDelegate != nullptr)
+    {
+        mAppDelegate->OnRendezvousStarted();
+    }
+
+    DeviceLayer::PlatformMgr().AddEventHandler(OnPlatformEventWrapper, reinterpret_cast<intptr_t>(this));
+
+    Cleanup();
+    ChipLogProgress(AppServer, "Device completed Rendezvous process");
+}
+
+CHIP_ERROR CommissioningWindowManager::PrepareCommissioningWindow(uint16_t commissioningTimeoutSeconds,
+                                                                  uint16_t & allocatedSessionID)
+{
+    ReturnErrorOnFailure(mIDAllocator->Allocate(allocatedSessionID));
+
+    ReturnErrorOnFailure(mPairingSession.MessageDispatch().Init(&mServer->GetSecureSessionManager()));
+
+    if (commissioningTimeoutSeconds != kNoCommissioningTimeout)
+    {
+        ReturnErrorOnFailure(
+            DeviceLayer::SystemLayer().StartTimer(commissioningTimeoutSeconds * 1000, HandleCommissioningWindowTimeout, this));
+    }
+
+    ReturnErrorOnFailure(mServer->GetExchangManager().RegisterUnsolicitedMessageHandlerForType(
+        Protocols::SecureChannel::MsgType::PBKDFParamRequest, &mPairingSession));
+
+    return StartAdvertisement();
+}
+
+CHIP_ERROR CommissioningWindowManager::OpenBasicCommissioningWindow(ResetFabrics resetFabrics, uint16_t commissioningTimeoutSeconds,
+                                                                    CommissioningWindowAdvertisement advertisementMode)
+{
+    RestoreDiscriminator();
+
+    // TODO: Verify that the device supports BLE if the advertisementMode is kBLE
+    // If not, then return an error.
+#if CONFIG_NETWORK_LAYER_BLE
+    SetBLE(advertisementMode == chip::CommissioningWindowAdvertisement::kBle);
+#else
+    SetBLE(false);
+#endif // CONFIG_NETWORK_LAYER_BLE
+
+    if (resetFabrics == ResetFabrics::kYes)
```

Comment:
What is the use case for resetting all fabrics when opening the commissioning window?  I would have expected this to only happen at factory reset.  The only command-initiated fabric removal I know if is the delete fabric command.

Reply:
@pan-apple ?

Reply:
The current m5stack implementation has a button UI where you can initiate a factory reset. This is just to keep the current use-case to work.

Reply:
How hard would it be to decouple this logic?  It seems like it would be cleaner for the m5stack's factory reset button to delete fabrics (and everything else), and then open the basic commissioning window.

It is surprising to me that fabric reset / deletion instead happens here, as this isn't a core function of this code.

Reply:
I'll check if that's something simple enough to integrate in this PR. If not, I'll file a TODO to do it in a follow on PR. The current approach was a placeholder added early on, while the fabric provisioning was not supported. I agree, it should be cleaned up now.